pkg/controller/toolchainstatus/toolchainstatus_controller.go

```diff
@@ -176,6 +184,35 @@ func (r *ReconcileToolchainStatus) aggregateAndUpdateStatus(reqLogger logr.Logge
 
 	// if any components were not ready then set the overall status to not ready
 	if len(unreadyComponents) > 0 {
+		// If the current ToolchainStatus:
+		// a) Is currently not ready,
+		// b) has not been ready for longer than the configured threshold, and
+		// c) no notification has been already sent, then
+		// send a notification to the admin mailing list
+		c, found := condition.FindConditionByType(toolchainStatus.Status.Conditions, toolchainv1alpha1.ConditionReady)
+		if found && c.Status == corev1.ConditionFalse {
+			threshold := time.Now().Add(-minutesAfterUnready * time.Minute)
+			if c.LastUpdatedTime.Before(&metav1.Time{threshold}) {
+				c, found = condition.FindConditionByType(toolchainStatus.Status.Conditions,
+					toolchainv1alpha1.ToolchainStatusUnreadyNotificationCreated)
+				if !found {
```

Comment:
Unless I'm missing something but this code will allow a single notification only. If it's ever sent once the notification condition is created and if later the status is going to the not-ready state again the notification will be skipped.
We need to reset/delete this condition when the status is ready again.
Also, just checking if the condition exists is not quite correct. Theoretically the condition can be in FALSE state.
So, it will be cleaner if we check that the condition is found and true.
Use https://github.com/codeready-toolchain/toolchain-common/blob/66539924b7b695797f8601631d69f4180ba2691d/pkg/condition/condition.go#L51 `condition.IsTrue()` instead.

Reply:
Remember that the `updateStatusConditions` function in this controller totally overwrites all existing conditions - keeping this in mind, when `setStatusReady()` is called it will as a result remove the notification condition:

```func (r *ReconcileToolchainStatus) setStatusReady(toolchainStatus *toolchainv1alpha1.ToolchainStatus) error {
	return r.updateStatusConditions(
		toolchainStatus,
		toolchainv1alpha1.Condition{
			Type:   toolchainv1alpha1.ConditionReady,
			Status: corev1.ConditionTrue,
			Reason: toolchainv1alpha1.ToolchainStatusAllComponentsReadyReason,
		})
}
```

Good point about using `condition.IsTrue()`, I'll switch to using it instead.

Reply:
Yeah I missed that. I would rename `updateStatusConditions()` to `replaceStatusConditions()` then. You are not changing this method here but it should be pretty trivial change.
We use `updateStatusConditions()` in almost every controller and it always updates or adds the given conditions and doesn't touch the existing ones which are not explicitly listed when this method is called. The only exception is our Toolchain (and probably member and host statuses too) Status controller. So, I would rename it to avoid confusions.

Reply:
Agreed, the function name is misleading.  I've renamed it.